File: DesignPatternDemo/Pattern/Decoration/DrcorationMain.m
//
//  DrcorationMain.m
//  DesignPatternDemo
//
//  Created by ruantong on 2019/5/11.
//  Copyright © 2019 wupeng. All rights reserved.
//

#import "DrcorationMain.h"
#import "FouthGradeSchoolReport.h"
#import "HighScoreDecorator.h"
#import "SortDecorator.h"

@implementation DrcorationMain

- (void)main{
    SchoolReport * sr = [FouthGradeSchoolReport new];
    sr = [[HighScoreDecorator alloc] initWithSr:sr];
    sr= [[SortDecorator alloc] initWithSr:sr];
    [sr report];
    [sr sign:@"老三"];
}

@end

File: DesignPatternDemo/Pattern/Decoration/DrcorationMain.swift
//
//  DrcorationMain.swift
//  DesignPatternDemo
//
//  Created by wupeng on 2020/3/4.
//  Copyright © 2020 wupeng. All rights reserved.
//

import Foundation

class DrcorationMain: NSObject,IMain {
    func main() {
        var sr:SchoolReport = FouthGradeSchoolReport();
        sr = HighScoreDecorator.init(sr: sr);
        sr = SortDecorator.init(sr: sr);
        sr.report()
        sr.sign(name: "老三")
    }
}
